lib/Sema/CSDiag.cpp

```diff
@@ -4575,6 +4576,22 @@ bool FailureDiagnosis::visitSubscriptExpr(SubscriptExpr *SE) {
 
         if (CS.TC.getTypeOfExpressionWithoutApplying(expr, CS.DC, decl))
           return false;
+
+        // The very specific, but maybe also fairly common(?) case where the subscript
+        // index type is an archetype of the base generic type. In which case we can
+        // get the real bound type and return a better error.
+        Type paramType = calleeInfo.candidates[0].getArgumentType();
+        if (auto archetype = paramType->getAs<ArchetypeType>()) {
```

Comment:
Take a look at how `CalleeCandidateInfo::evaluateCloseness` uses `findGenericSubstitutions`, I think it would be nice if this check could do something like - if argument is `UnresolvedMemberExpr` and parameter type is generic and closeness is near miss, use this generic type as argument type to check if it is valid by calling `typeCheckChildIndendently` like what you do on L4592.

Reply:
Not quite the same thing, since ```findGenericSubstitutions``` is trying to find shared substitutions between between multiple arguments that involve the same generic params. But I did work out the more general correct code, thanks!